### PR TITLE
Fixed hot start gradient bug

### PR DESCRIPTION
**Pull request by sseraj**
## Purpose

If an optimization crashes in the middle of a gradient call, attempting to hot start the optimization will give the following error:

```
.../pyoptsparse/pyOpt_optimizer.py", line 454, in _masterFunc2
    args = self.sens(xuser, self.cache["funcs"])
KeyError: 'funcs'
```

I fixed this by re-evaluating the functions if the `funcs` key does not exist.

## Expected time until merged

Regular urgency; one week

## Type of change

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing

I tried to create a test, but this bug proved difficult to reproduce even with manual modification of the history file.

## Checklist

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

### CODE REVIEW

**Comment by sseraj** [comment]
I'm guessing we will merge #284 first, which will fix the flake8 issue.

**Comment by marcomangano** [comment]
This is looking good, but I am surprised that the manual modification of the `.hst` file does not work. If you just remove the last entry, then the hotstart should break. We can talk offline about this - I had a script that pops out dict entries made specifically to circumvent this issue :)

File: pyoptsparse/pyOpt_optimizer.py [hunk 1]
```diff
@@ -438,10 +444,10 @@ def _masterFunc2(self, x, evaluate, writeHist=True):
...
         if "gobj" in evaluate:
-            if not np.isclose(x, self.cache["x"], atol=EPS, rtol=EPS).all():
-                # Previous evaluated point is *different* than the
-                # point requested for the derivative. Recursively call
-                # the routine with ['fobj', and 'fcon']
+            if not np.isclose(x, self.cache["x"], atol=EPS, rtol=EPS).all() or "funcs" not in self.cache:
+                # The previous evaluated point is different than the point requested for the derivative
+                # OR this is the first call to _masterFunc2 in a hot started optimization
+                # Recursively call the routine with ['fobj', 'fcon']
                 self._masterFunc2(x, ["fobj", "fcon"], writeHist=False)
```
[the remark this answers is not in the document]
Comment (marcomangano) [reply pyoptsparse/pyOpt_optimizer.py:451]: yeah we should address that in a separate PR

**Review by nwu63 — APPROVED** [review]
LGTM

**Review by marcomangano — APPROVED** [review]

**Pull request merged**